File: 50dias/python/43.py
import os
votacao = {19: {'nome': 'Leonardo',
                'quantidade': 0},
           27: {'nome': 'Luizinho',
                'quantidade': 0},
           33: {'nome': 'Alfredo',
                'quantidade': 0},
           42: {'nome': 'Eymael',
                'quantidade': 0},
           500: {'nome': 'Branco',
                 'quantidade': 0}}

votosNulos = 0
votosTotais = 0

def ImprimeMenu():
    print('---------------------------------------------------')
    print('        Sistema de eleições presidenciais          ')
    print('---------------------------------------------------')
    print('Vote em um dos candidatos: ')
    print('   19 - Leonardo')
    print('   27 - Luizinho')
    print('   33 - Alfredo')
    print('   42 - Eymael')
    print('  500 - Voto em Branco')
    print('---------------------------------------------------')

def ConfirmaVoto(numeroVotado):
    if numeroVotado in votacao.keys():
        nome = votacao[numeroVotado]['nome']
        print('O seu voto é no candidato: {}'.format(nome))
    else:
        print('Candidato invalido!')
    confirmacao = input('Deseja confirmar? [s]im/[n]ao: ')
    return confirmacao == 's'

def ContabilizaVoto(numeroVotado):
    global votosTotais, votosNulos
    votosTotais += 1
    if numeroVotado in votacao.keys():
        votacao[numeroVotado]['quantidade'] += 1
    else:
        votosNulos += 1

def ImprimeRelatorioEleicao():
    if votosTotais > 0:
        votosBrancos = votacao[500]['quantidade']
        votosInvalidos = votosBrancos + votosNulos
        votosValidos = votosTotais - votosInvalidos
        os.system("cls")
        print("--------------------------------------")
        print("     Votos validos por candidato")
        print("--------------------------------------")
        for numero, candidato in votacao.items():
            if numero != 500:
                nome = candidato['nome']
                quant = candidato['quantidade']
                porc = quant*100/votosValidos
                msg = "{nome} - {votos} - {porc:.2f}%".format(nome=nome,
                                                              votos=quant,
                                                              porc=porc)
                print(msg)

        porcBrancos = votosBrancos*100/votosTotais
        porcNulos = votosNulos*100/votosTotais
        print("--------------------------------------")
        print('Votos Brancos: {quant} - {porc:.2f}%'.format(quant=votosBrancos,
                                                           porc=porcBrancos))
        print('Votos Nulos: {quant} - {porc:.2f}%'.format(quant=votosNulos,
                                                     porc=porcNulos))

def LeVotoEleitor():
    votacaoEmAndamento = True
    numeroVotado = 0
    while votacaoEmAndamento:
        os.system("cls")
        ImprimeMenu()
        numeroVotado = int(input('Digite o numero do candidato: '))
        votoConfirmado = ConfirmaVoto(numeroVotado) if numeroVotado != 0 else True
        votacaoEmAndamento = not votoConfirmado
    return numeroVotado



def main():
    continuaVotacao = True
    while continuaVotacao:
        numeroVotado = LeVotoEleitor()
        continuaVotacao = numeroVotado != 0
        if continuaVotacao:
            ContabilizaVoto(numeroVotado)
            print('Voto contabilizado!')
            input('Pressione ENTER para finalizar!')

    ImprimeRelatorioEleicao()

main()

File: 50dias/cpp/43.cpp
#include <iostream>
#include <map>
#include <iomanip>
#include <conio.h>

using namespace std;

map<int, string> candidatos = { {19, "Leonardo"},
                                {27, "Luizinho"},
                                {33, "Alfredo"},
                                {42, "Eymael"},
                                {500, "Branco"} };
map<int, int> votacao = { {19, 0},
                          {27, 0},
                          {33, 0},
                          {42, 0},
                          {500, 0} };

int votosNulos = 0;
int votosTotais = 0;

void ImprimeMenu() {
    cout << "---------------------------------------------------" << endl;
    cout << "        Sistema de eleicoes presidenciais          " << endl;
    cout << "---------------------------------------------------" << endl;
    cout << "Vote em um dos candidatos: " << endl;
    cout << "   19 - Leonardo" << endl;
    cout << "   27 - Luizinho" << endl;
    cout << "   33 - Alfredo" << endl;
    cout << "   42 - Eymael" << endl;
    cout << "  500 - Voto em Branco" << endl;
    cout << "---------------------------------------------------" << endl;
}

bool ConfirmaVoto(int numeroVotado) {
    if (candidatos.find(numeroVotado) != candidatos.end()) {
        cout << "O seu voto e no candidato ";
        cout << candidatos[numeroVotado] << endl;
    }
    else {
        cout << "Candidato invalido!" << endl;
    }
    string confirmacao;
    cout << "Deseja confirmar? [s]im/[n]ao: ";
    cin >> confirmacao;
    return confirmacao == "s";
}

int LeVotoEleitor() {
    int numeroVotado = 0;
    bool votacaoEmAndamento = false;
    do {
        system("cls");
        ImprimeMenu();
        cout << "Digite o numero do candidato: ";
        cin >> numeroVotado;
        bool votoConfirmado = numeroVotado != 0 ? ConfirmaVoto(numeroVotado) : true;
        votacaoEmAndamento = !votoConfirmado;
    } while (votacaoEmAndamento);
    return numeroVotado;
}

void ContabilizaVoto(int numeroVotado) {
    votosTotais++;
    if (votacao.find(numeroVotado) != votacao.end()) {
        votacao[numeroVotado]++;
    }
    else {
        votosNulos++;
    }
}

void ImprimeRelatorioEleicao() {
    if (votosTotais != 0) {
        int votosBrancos = votacao[500];
        int votosInvalidos = votosBrancos + votosNulos;
        int votosValidos = votosTotais - votosInvalidos;
        system("cls");
        cout << "--------------------------------------" << endl;
        cout << "       Votos validos por candidato    " << endl;
        cout << "--------------------------------------" << endl;

        for (auto item : votacao) {
            int numero = item.first;
            if (numero != 500) {
                string nome = candidatos[numero];
                int quant = votacao[numero];
                int porc = float(quant * 100) / votosValidos;
                cout << nome << " - Votos: " << quant;
                cout << " : " << fixed << setprecision(2);
                cout << porc << "%" << endl;
            }
        }

        float porcBranco = float(votosBrancos * 100) / votosTotais;
        float porcNulos = float(votosNulos * 100) / votosTotais;
        cout << "--------------------------------------" << endl;
        cout << "Votos Brancos: " << votosBrancos;
        cout << " - " << porcBranco << "%" << endl;
        cout << "Votos Nulos: " << votosNulos;
        cout << " - " << porcNulos << "%" << endl;
    }
}

int main()
{
    bool continuaVotacao = false;
    do {
        int numeroVotado = LeVotoEleitor();
        continuaVotacao = numeroVotado != 0;
        if (continuaVotacao) {
            ContabilizaVoto(numeroVotado);
            cout << "Voto contabilizado!" << endl;
            cout << "Pressione Enter tecla para continuar!";
            getch();
        }
    } while (continuaVotacao);
    ImprimeRelatorioEleicao();

    return 0;
}